### DISCUSSION TITLE
Installation on Ubuntu 23.04

### DISCUSSION THREAD

**Discussion by sojusnik** [facelessuser/Rummage#452] [Q&A] [answered]
Hey!

According to your [instructions,](https://facelessuser.github.io/Rummage/installation/) it should be enough to install a pre-build distro package of `wxPython` and then install Rummage by using `pip`. Since I'm using Ubuntu, the `pip` command isn't working any more and has to be replaced by `pipx`, as described [here](https://www.omgubuntu.co.uk/2023/04/pip-install-error-externally-managed-environment-fix).

So for `wxPython` I installed `python3-wxgtk4.0` (4.2.0), but after `pipx install rummage` I only see a spinning animation showing that the installation is proceeding, but because it didn't stop after roughly 15 minutes, I had to cancel the installation. No error message or the like was shown.

Any ideas what I'm doing wrong?

**Comment by facelessuser** [comment]
If it was taking a long time, it was likely building wxpython as they don't distribute wheels for Linux. You can manually download prebuilt wheels though for Ubuntu. You'll have to wait until I'm at my computer for more info.

**Comment by facelessuser** [comment]
Pre-built wheels are here: https://extras.wxpython.org/wxPython4/extras/linux/gtk3/.

I don't know if `python3-wxgtk4.0` includes wxpython or just some related tools. If you are install wxpython via pip, it can take a while to build the whole project, and it won't give you any indications until it either finishes or fails due to some missing dependency or build error.

I'd have to spin up a virtual machine with Ubuntu 23.04 to see exactly what the requirements are. Linux is always a moving target, and wxPython doesn't distribute the wheels for the many Linux variations.

**Comment by sojusnik** [comment]
> Pre-built wheels are here: https://extras.wxpython.org/wxPython4/extras/linux/gtk3/.

It seems that I can't use them with `pipx`, as I've tried different combinations, but it didn't start to download it.

Isn't it so that packages installed from the [distro's repository](https://packages.ubuntu.com/lunar/python3-wxgtk4.0) don't have to be built? According to [here](https://askubuntu.com/a/1154318/79691), installing `python3-wxgtk4.0` should be enough to have `wxPython` installed, doesn't it?

So I assume that the endless installation process doesn't have to do with `wxPython`.

**Comment by facelessuser** [comment]
Right now you've provided so few details that is difficult for me to guess what your issue is.

1. Have you tried downloading the Linux wheel to your machine and then and then having pipx process it? I really don't know what "I've tried different combinations" means. I don't currently use `pipx`, so I don't know what it can and cannot do.
2. If `wxgtk4.0` does install all of wxpython, are you sure that `pipx` is installing for the default python on Ubuntu? I've seen issues where people have multiple versions of python on their machine and they think they are installing rummage for one version but accidentally install it for another. Or maybe installing it in one venv environment that doesn't have the dependencies they think it does and so it then has to install wxpython (or something else) again. So yes, it is possible that even if you have `wxgtk4.0` that it is trying to install wxpython, but I don't have enough details from you to determine this.
4. Rummage is a simple, pure python wheel, so if it is hanging, it is hanging on some other dependency, it is one that it has to build. Pure Python wheels are just unpacked to the right place, they are not built.
5. Does `pipx` have a verbose mode? Maybe run with that and see what it is getting stuck on. It should be able to list what the various package dependencies it is attempting to install are and maybe even the step in the process it is on.

Without more details, all I can do is attempt to install it on a fresh Ubuntu virtual machine and see if there are any extra steps that are required due to missing dependencies. Even then, if there are things different in your environment, there could be gaps. I'm sorry that Linux installation is a pain, it has always been that way, unfortunately. The hardest part is always ensuring the basic dependencies are there. Once they are in place, wxpython being the biggest hurdle, things usually go smooth and quick.

I can probably try installing this on an Ubuntu VM over the weekend and can note what I had to do to get it installed properly.

**Comment by sojusnik** [comment]
Hey!

1. By “different combinations” with `pipx` I mean something like this:

```
$ pipx install -U -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-22.04 wxPython
usage: pipx [-h] [--version]
            {install,inject,upgrade,upgrade-all,uninstall,uninstall-all,reinstall,reinstall-all,list,run,runpip,ensurepath,environment,completions}
            ...
pipx: error: unrecognized arguments: -U wxPython

$ pipx install -f https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-22.04 wxPython
usage: pipx [-h] [--version]
            {install,inject,upgrade,upgrade-all,uninstall,uninstall-all,reinstall,reinstall-all,list,run,runpip,ensurepath,environment,completions}
            ...
pipx: error: unrecognized arguments: wxPython

$ pipx install https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-22.04 wxPython
usage: pipx [-h] [--version]
            {install,inject,upgrade,upgrade-all,uninstall,uninstall-all,reinstall,reinstall-all,list,run,runpip,ensurepath,environment,completions}
            ...
pipx: error: unrecognized arguments: wxPython

$ pipx install https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-22.04
  ERROR: Cannot unpack file /tmp/pip-unpack-upqzeo6l/ubuntu-22.04 (downloaded from /tmp/pip-req-build-kw45u8ok, content-type: text/html;charset=UTF-8); cannot detect archive format
ERROR: Cannot determine archive format of /tmp/pip-req-build-kw45u8ok
Cannot determine package name from spec
'https://extras.wxpython.org/wxPython4/extras/linux/gtk3/ubuntu-22.04'. Check package
spec for errors.
```

2. I'm not sure at all, just saw that somebody recommended installing it.

4. None that I know of, but that doesn't mean much, because I'm not into coding and don't have practical experience with python.

**Comment by facelessuser** [comment]
I think you would have to give it the full path to the wheel `.whl` file, not the folder.  Anyway, I'll have to find some time to do a test on installing Rummage on the latest Ubuntu. I know they change things every time, and they now have global install restrictions as well.

**Comment by facelessuser** [comment, marked as answer]
I've installed this on a fresh Ubuntu 23.04 in virtual box.

1. I installed `python3-wxgtk4.0`
2. I also had to install `python3-wxgtk-webview4.0`
3. I had to install rummage with access to these system wide packages as they are not available via the virtual environments that `pipx` creates by default: `pipx install --system-site-packages rummage`

Then I could access `rummage`:

<img width="1232" alt="Screenshot 2023-12-31 at 12 30 08 PM" src="https://github.com/facelessuser/Rummage/assets/1055125/5fd50d3d-2dae-4acf-a532-0bb95f168d2f">

**Reply by sojusnik** [reply]
Can confirm that it also works on 23.10!

For a fresh installation:

`sudo apt install python3-wxgtk4.0 python3-wxgtk-webview4.0`
`pipx install --system-site-packages rummage`

or if you want to upgrade your package (when you have backuped `/home/yourusername/.local/pipx` and then restored it), then:

`sudo apt install python3-wxgtk4.0 python3-wxgtk-webview4.0`
`pipx upgrade --system-site-packages rummage`

BTW, you have a small (but fatal) typo in your [instruction](https://facelessuser.github.io/Rummage/installation/#pipx), where you have written

`sudo apt install python3-wgtk4.0 python3-wxgtk-webview4.0`
instead of
`sudo apt install python3-wxgtk4.0 python3-wxgtk-webview4.0`

Keep up the good work!

**Reply by facelessuser** [reply]
Thanks for letting me know about the typo, I'll get that updated.

**Comment by facelessuser** [comment]
Moral of the story is that Ubuntu has made this even more unintuitive with their recent Python restrictions.

Pipx is great at creating an "app", but each time it creates an "app", it does a completely clean environment and only installs what you ask it to, so it doesn't know wxpython was installed via Ubuntu's own curated packages. To get around that, you have to tell pipx that it needs to include such when creating your app. 

I will likely create a new section in the docs describing this.

Sorry it took me a bit to answer this. I was pretty busy at work, then got pretty sick.

**Reply by sojusnik** [reply]
Thanks for your commitment! Following your instructions I was able to install it on my machine as well! May you have an awesome 2024!

**Reply by sojusnik** [reply]
Just to be sure: Is `pipx upgrade --system-site-packages rummage` the right way to update?

I've used this command and it seems to work well.

**Reply by facelessuser** [reply]
I guess. I've only used pipx to check installation. I had never used it before. But if it seemed to work, I imagine that must be the way.
